[CPU] Add an OpenMP-based CPU launcher

minjang opened the pull request:
This PR implements a simple OpenMP-based CPU launcher with some debugging and tunable environment variables. 

**Details:**
- By default, it uses the entire threads on a machine. So there might be a chance of contention between logical threads on the same physical core. Need a fine tuning.
- It uses static scheduling. No reason to use dynamic scheduling for now (overhead is high). I used the default chunk size: total iterations / # of threads, where the minimum chunk size is 10.

**Initial performance:**

Environment: I only tested on an AMD EPYC Zen4 96-core machine with Centos 9, as a development server in Meta. My test wasn't on a physical machine; it is a VM.

I only tested the vector addition example. I added `TritonCPU 1-core` as well as `TorchCPU` as the references. I disabled the turbo boost-like feature during this test. I used a smaller `BLOCK_SIZE` of 128 for CPU. Some observations so far:
- For large vector sizes >= 2^24, the GB/s seems to be stable.
  - The performance of `TritonCPU` with maximum threads is similar to that of `TorchCPU`.
- A large variance is observed in the sizes of 2^18 to 2^22, for both `TritonCPU` and `TorchCPU`
- The OpenMP's speedups for larger sizes are not impressive: ~6x.

```
> % python3 python/tutorials/01-vector-add.py
tensor([0.5151, 1.6826, 0.9153,  ..., 0.9852, 1.2714, 1.8192])
tensor([0.5151, 1.6826, 0.9153,  ..., 0.9852, 1.2714, 1.8192])
The maximum difference between torch-cpu and triton-cpu is 0.0
tensor([0.5151, 1.6826, 0.9153,  ..., 0.9852, 1.2714, 1.8192], device='cuda:0')
tensor([0.5151, 1.6826, 0.9153,  ..., 0.9852, 1.2714, 1.8192], device='cuda:0')
The maximum difference between torch-gpu and triton-gpu is 0.0
vector-add-performance (CPU_BLOCK_SIZE=128, GPU_BLOCK_SIZE=1024):
           size  TritonCPU 1-core  TritonCPU    TorchCPU    TritonGPU     TorchGPU
0        4096.0          0.403202   0.062194    4.472021    10.816901    10.666666
1        8192.0          3.120957   1.619745    7.883864    21.186207    21.041095
2       16384.0          3.280629   0.958061    3.545745    41.234899    40.959998
3       32768.0          3.188169   2.692145    6.301640    81.919996    80.313725
4       65536.0          3.969092   0.291063    7.116936   158.554837   156.535030
5      131072.0          5.689773   5.975995    9.832181   284.115618   282.482757
6      262144.0          5.911389  12.657541    4.723294   484.256150   489.074621
7      524288.0          6.442556  15.187768    5.880182   783.298835   771.011790
8     1048576.0          6.401166  55.634510    1.416960  1153.126078  1149.754339
9     2097152.0         10.465727  32.779651   98.862590  1521.145131  1481.039473
10    4194304.0          9.414268  34.194252  195.153484  1767.262872  1745.687030
11    8388608.0          3.347814   4.684119    5.160477  1972.861670  1957.515852
12   16777216.0          3.677767   7.394172    7.372241  2109.103660  2077.759573
13   33554432.0          3.057506  12.829415   10.310923  2168.532851  2172.839199
14   67108864.0          3.555315  17.150573   15.640461  2226.669987  2237.160951
15  134217728.0          3.711412  16.852324   18.304376  2246.948652  2260.369439
```

**TODO:**
- More experimentation with the fused softmax case.
- More rigorous performance testing on a stable environment.

**Debugging and control features**

`TRITON_CPU_MAX_THREADS` and `TRITON_CPU_OMP_DEBUG` are supported.

```
TRITON_CPU_MAX_THREADS=40 TRITON_CPU_OMP_DEBUG=1 python3 python/tutorials/01-vector-add.py
```

ienkovich commented:
Thanks for the patch, looks like a good start!

Some results look suspicious though. There is a big perf drop starting from 8MB size. The drop happens for all CPU options. I tried it on my machine, it shows quite stable perf growth with size increase but it also has this drop at 8MB:
```
           size  TritonCPU 1-core   TritonCPU    TorchCPU
0        4096.0          0.473994    0.369581    1.530837
1        8192.0          0.790753    0.747536    2.880876
2       16384.0          1.479289    1.597310    4.705235
3       32768.0          2.556686    3.840678    9.008160
4       65536.0          5.013719    8.386279   12.160891
5      131072.0          7.338396   14.899770   25.372058
6      262144.0         10.075180   29.821873   51.502574
7      524288.0         11.859673   36.852700   91.065672
8     1048576.0         12.967142   61.431152  110.480621
9     2097152.0         13.879187   87.099157  172.930611
10    4194304.0         14.490595  101.303109  220.705400
11    8388608.0          6.876308   23.051736   26.772240
12   16777216.0          6.798736   27.438252   24.234607
13   33554432.0          6.649066   25.322378   26.525232
14   67108864.0          6.489616   28.934346   30.688774
15  134217728.0          6.335913   28.097894   29.345317
```
Do you have any theory why it might happen?

ienkovich commented on third_party/cpu/backend/driver.py:
```diff
@@ -16,6 +16,7 @@
 include_dir = [
     os.path.join(dirname, "include"),
     os.path.join(llvm_root, "include"),
+    os.path.join(".", "include"),
```
Does it mean Triton can now be run only from the root repo dir?

minjang replied:
This was to use `GetEnv.hpp`. I will remove

ienkovich commented on third_party/cpu/backend/driver.py:
```diff
@@ -77,6 +78,8 @@
     "z",
 ]
 
+MINIMUM_OMP_CHUNK_SIZE = 10
```
Can you please explain why we need this one? I can easily imagine cases when a chunk size of 1 is preferred.

minjang replied:
My thought was that a typical iteration in the omp-parallel-for loop in TritonCPU would tend to be (1) uniform and have (2) small amount of the workload. So, having a smallest chunk size like 1 might have more scheduling overhead, losing cache affinity as well as potential prefetch.

Let me rethink on how to control chunk size.

ienkovich replied:
It's hard to come up with a universal heuristic here. For vector-add, we have a very small block processed by a kernel, and setting the minimum chunk size to 10 looks reasonable. But for matmul kernels that would be very different. We might have dozens of quite heavy kernels and this minimum chunk size would significantly limit parallelism.

We definitely cannot find the best strategy based on the vector-add example only and will have multiple iterations here. So it's OK to start with some basics for now, but I'd like to avoid any hard-coded limits here that cannot be controlled in our perf experiments.

minjang replied:
Right. This was more like an interim testing code. For now, let's just use the default chunk size, that is, don't specify it. I removed the code that overrides chunk size.

When we have more perf benchmarks, we can introduce some good overridable options.

ienkovich commented on third_party/cpu/backend/driver.py:
```diff
@@ -186,6 +188,10 @@ def format_of(ty):
+#include <omp.h>
+#include <cmath>
+
+#include "triton/Tools/Sys/GetEnv.hpp"
```
I don't think `GetEnv.hpp` is a part of the Triton package. This code is compiled on a user's machine and therefore header would be unavailable.

minjang replied:
Oh right. Let me just implement simple helper functions directly in here.

ienkovich commented on third_party/cpu/backend/driver.py:
```diff
@@ -233,20 +239,56 @@ def format_of(ty):
 
-static PyObject* launch(PyObject* self, PyObject* args) {{
+static void run_omp_kernels(uint32_t gridX, uint32_t gridY, uint32_t gridZ, kernel_ptr_t kernel_ptr {', ' + arg_decls if len(arg_decls) > 0 else ''}) {{
+  auto all_grids = get_all_grids(gridX, gridY, gridZ);
```
Can we try `collapse(3)` OMP rule instead of building these grids?

minjang replied:
That's good idea. But I had some troubles to get it right with clang compiler with OpenMP. E.g., only one cpu was used etc. So, my suggestion is that keep the current approach for the time being. Once we confirm MacOS+Clang works well with OpenMP, then I will try `collapse` clause to make the code simpler.

ienkovich commented on third_party/cpu/backend/driver.py:
```diff
@@ -233,20 +239,56 @@ def format_of(ty):
+    max_threads = omp_get_max_threads();
 
+  int chunk_size = std::ceil((double)N / (double)max_threads.value());
+  chunk_size = std::max(chunk_size, {MINIMUM_OMP_CHUNK_SIZE});
```
The chunk size can affect performance a lot. For our experiments, I'd remove the min border and add an opportunity to override the chunk size instead.

ienkovich commented on third_party/cpu/backend/driver.py:
```diff
@@ -233,20 +239,56 @@ def format_of(ty):
+  chunk_size = std::max(chunk_size, {MINIMUM_OMP_CHUNK_SIZE});
+
+  if (mlir::triton::tools::getBoolEnv("TRITON_CPU_OMP_DEBUG"))
+    printf("N: %zu, max_threads: %d, chunk_size: %zu\\n", N, max_threads, chunk_size);
```
You pass `std::optional<int> max_threads` here instead of `int` value.

ienkovich replied:
`chunk_size` type doesn't match the specifier `%zu`

ienkovich commented on third_party/cpu/backend/driver.py:
```diff
@@ -233,20 +239,56 @@ def format_of(ty):
+  if (mlir::triton::tools::getBoolEnv("TRITON_CPU_OMP_DEBUG"))
+    printf("N: %zu, max_threads: %d, chunk_size: %zu\\n", N, max_threads, chunk_size);
+
+#pragma omp parallel for schedule(static, chunk_size) num_threads(max_threads.value())
```
I get much better results if remove `chunk_size` from here. Can you try such an option?

minjang replied:
If no chunk_size, the default behavior should be the same in my code (unless the minimum chunk), which is total iteration / # of threads. Let me do some update for the chunk size.

minjang commented:
> 6.876308   23.051736   26.772240
> 12   16777216.0          6.798736   27.438252   24.234607
> 13   33554432.0          6.649066   25.322378   26.525232
> 14   67108864.0          6.489616   28.934346   30.688774
> 15  134217728.0          6.335913   28.097894   29.345317

Your numbers look much stable than me. Note that I was doing on our devvm. This big drop at 8MB size is confusing us as well. Our engineers will also take a look at.

embg commented:
> This big drop at 8MB size is confusing us as well.

In the past, when I have hit weird benchmark issues at 8MB, it was because of jemalloc. It caches allocations smaller than 8MB, but not larger.

Is it possible to rewrite the benchmark so that it avoids malloc / free, and uses the same memory each time?

minjang commented:
> Edit: never mind, the drop is at 8M _elements_, not 8M _bytes_, disregard my comment.

Nonetheless, the memory allocator still might be a reason. This is a good pointer.

I'm not sure whether I made a terrible mistake in `do_bench` with `is_cpu=True` in `python/triton/testing.py`. Originally, `do_bench` only considered GPU. So I patched `do_bench` for CPU. It does allocate some chunk of memory to flush L3 cache.

Let's do some more investigation.

embg commented:
> It does allocate some chunk of memory to flush L3 cache.

It would be better if we can flush L3 cache by writing to an existing allocation. So we don't have page faults while flushing L3 cache.

ienkovich commented:
> Nonetheless, the memory allocator still might be a reason. This is a good pointer.

Moving the creation of the output tensor out of the `add` functions 'fixes' the problem. So the problem seems to be in this allocation. Maybe the torch initializes tensors with zeros for big sizes or a garbage collection takes place here?

minjang commented:
> > Nonetheless, the memory allocator still might be a reason. This is a good pointer.
> 
> Moving the creation of the output tensor out of the `add` functions 'fixes' the problem. So the problem seems to be in this allocation. Maybe the torch initializes tensors with zeros for big sizes or a garbage collection takes place here?

Thanks for the suggestion! Let me "fix" it as well.

minjang commented:
Updated:
- Removed chunk size override. For now, it's okay to use the default chunk size policy.
- Introduced "prealloc" of `output` for TritonCPU in the vector-add benchmark.
- Removed the dependency of using `GetEnv.hpp`.

minjang commented:
It turned out `torch.add` can take preallocated output buffer. Let's simply always have preallocated output buffer for CPU cases.

```
> % python3 python/tutorials/01-vector-add.py
tensor([0.5151, 1.6826, 0.9153,  ..., 0.9852, 1.2714, 1.8192])
tensor([0.5151, 1.6826, 0.9153,  ..., 0.9852, 1.2714, 1.8192])
The maximum difference between torch-cpu and triton-cpu is 0.0
tensor([0.5151, 1.6826, 0.9153,  ..., 0.9852, 1.2714, 1.8192], device='cuda:0')
tensor([0.5151, 1.6826, 0.9153,  ..., 0.9852, 1.2714, 1.8192], device='cuda:0')
The maximum difference between torch-gpu and triton-gpu is 0.0
vector-add-performance (CPU_BLOCK_SIZE=4096, GPU_BLOCK_SIZE=1024):
           size  TritonCPU 1   TritonCPU    TorchCPU    TritonGPU     TorchGPU
0        4096.0     0.330691    0.003187    1.640316    11.050360    10.893617
1        8192.0     0.481347    0.011473    2.034690    21.787235    21.633802
2       16384.0     1.036711    0.012717    3.091976    42.372414    42.082190
3       32768.0     2.670090    0.025314   11.095744    83.591837    82.469799
4       65536.0     3.189345    0.048478    0.040158   161.684218   160.627450
5      131072.0    10.870723    9.019075    0.094202   290.840226   287.438585
6      262144.0     8.618670    7.680253   27.162953   493.989945   499.005073
7      524288.0     5.992274   25.221713   33.713298   792.774204   780.190482
8     1048576.0     6.782821   89.671679   65.595797  1166.813039  1159.929234
9     2097152.0     7.074116   68.878399    1.539270  1533.005829  1492.280845
10    4194304.0     5.453523  140.154074  237.239931  1773.240131  1755.428547
11    8388608.0     5.936637  226.831542  267.304201  1977.201751  1963.625499
12   16777216.0     5.669703   12.547080  235.037567  2108.396850  2077.759573
13   33554432.0     5.743683  165.028759  218.124103  2170.590254  2174.716971
14   67108864.0     7.190059  129.821671  127.623851  2227.458335  2237.558870
15  134217728.0     7.682400   70.595290   77.931064  2246.697937  2260.978822
```

ienkovich reviewed (approved):
Looks good! Thanks!
